streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala

```diff
@@ -298,6 +298,41 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
                         stateStoreNames: String*): KStream[K1, V1] =
     inner.transform(transformerSupplier, stateStoreNames: _*)
 
+
+  /**
+   * Transform each record of the input stream into zero or more records in the output stream (both key and value type
+   * can be altered arbitrarily).
+   * A `Transformer` (provided by the given `TransformerSupplier`) is applied to each input record
+   * and computes zero or more output records.
+   * In order to assign a state, the state must be created and registered
+   * beforehand via stores added via `addStateStore` or `addGlobalStore`
+   * before they can be connected to the `Transformer`
+   *
+   * @param transformerSupplier the `TransformerSuplier` that generates `Transformer` (this function variant expects a transformer that returns a (key, value) tuple)
+   * @param stateStoreNames     the names of the state stores used by the processor
+   * @return a [[KStream]] that contains more or less records with new key and value (possibly of different type)
+   * @see `org.apache.kafka.streams.kstream.KStream#transform`
+   */
+  def transform[K1, V1](transformerSupplier: () => Transformer[K, V, (K1, V1)],
+                        stateStoreNames: String*): KStream[K1, V1] = {
+    val transformerSupplierJ = new TransformerSupplier[K, V, KeyValue[K1, V1]] {
+      override def get(): Transformer[K, V, KeyValue[K1, V1]] = {
+        val scalaTransformer = transformerSupplier()
+        new Transformer[K, V, KeyValue[K1, V1]] {
+          override def init(context: ProcessorContext): Unit = scalaTransformer.init(context)
+          override def transform(key: K, value: V): KeyValue[K1, V1] = {
+            scalaTransformer.transform(key, value) match {
+              case (k1, v1) => KeyValue.pair(k1, v1)
+              case _ => null
```

Comment:
Am I right in thinking that the only other permissible return value is `null`? Do you mind explaining the wildcard case?

Reply:
I really think that the only value that doesn't match (k1, v1) can be null. So I don't mind changing it to null => null. I used a wildcard because that was the way the code was written in 2.0.0 version.